src/Tools/ExternalAccess/FSharp/Internal/Editor/FSharpSynchronousIndentationService.cs

```diff
@@ -2,42 +2,65 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-#nullable disable
-
 using System;
 using System.Composition;
 using System.Threading;
 using Microsoft.CodeAnalysis.ExternalAccess.FSharp.Editor;
 using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.Host.Mef;
 using Microsoft.CodeAnalysis.Indentation;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.ExternalAccess.FSharp.Internal.Editor
 {
     [Shared]
     [ExportLanguageService(typeof(IIndentationService), LanguageNames.FSharp)]
     internal class FSharpSynchronousIndentationService : IIndentationService
     {
-        private readonly IFSharpSynchronousIndentationService _service;
+#pragma warning disable CS0618 // Type or member is obsolete
+        private readonly IFSharpSynchronousIndentationService? _legacyService;
+#pragma warning restore
+        private readonly IFSharpIndentationService? _service;
 
         [ImportingConstructor]
         [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
-        public FSharpSynchronousIndentationService(IFSharpSynchronousIndentationService service)
+        public FSharpSynchronousIndentationService(
+            [Import(AllowDefault = true)] IFSharpSynchronousIndentationService? legacyService,
+            [Import(AllowDefault = true)] IFSharpIndentationService? service)
         {
+            Contract.ThrowIfTrue(service == null && legacyService == null);
+
+            _legacyService = legacyService;
             _service = service;
         }
 
         public IndentationResult GetIndentation(Document document, int lineNumber, FormattingOptions.IndentStyle indentStyle, CancellationToken cancellationToken)
         {
-            var result = _service.GetDesiredIndentation(document, lineNumber, cancellationToken);
-            if (result.HasValue)
+            // all F# documents should have a file path
+            if (document.FilePath == null)
+            {
+                return default;
+            }
+
+            FSharpIndentationResult? result;
+            if (_service != null)
             {
-                return new IndentationResult(result.Value.BasePosition, result.Value.Offset);
+                var text = document.GetTextSynchronously(cancellationToken);
+                var documentOptions = document.GetOptionsAsync(cancellationToken).WaitAndGetResult(cancellationToken);
```

Comment:
@jasonmalinowski any way to not have to be async+blocking just to find out tab size?

Reply:
Currently we can't get the necessary options from the editor. Working on it with the editor team.